README.md

```diff
@@ -66,19 +70,97 @@ class EmailAddressMatches implements Specification
+
+The comparison operator is `is`, which means "equals". You can use `ComparisonNode::isEquals()`, 
+`ComparisonNode::isLessThan()` etc to determine the comparison type.
+
+Finally, `karl.rixon@gmail.com` is converted into a `StringNode` AST node during parsing. This node can be accessed
+via `ComparisonNode::value()`.
+
+Based on the above, the `BaseCompiler::generate()` method might be implemented as follows:
 
 ```php
-public function literal(IdentifierNode $identifier, LiteralNode $node) : Specification
+class MyCompiler extends BaseCompiler
 {
-    switch (strtolower($identifier->fullName())) {
-        case 'email':
-            return new EmailAddressMatches($node->value());
-        // case ...
-    }
-
-    throw new CompilerError(sprintf("Unknown identifier '%s'.", $identifier->fullName()));
+  public function generate(ComparisonNode $comparison) : Specification
+  {
+      $identifier = $comparison->identifierFullName();
+      
+      if (strtolower($indentifier) !== 'email') {
+          throw CompilerError::unknownIdentifier($identifier);
+      }
+      
+      if (!$comparison->isEquals()) {
+          throw CompilerError::unsupportedComparisonType($comparison->type(), $identifier);
+      }
+  
+      return new EmailAddressMatches($comparison->literalValue());
+  }
 }
+```
+
+## Delegating generation to services
+
+Although extending `BaseCompiler` is convenient in simple cases, it becomes complicated when you have many
+specifications to support. In this case, you might want to delegate the generation work to dedicated services.
+
+The `DelegatingCompiler` class is provided for this purpose. To use it, first create a class which implements the
+`SpecificationGenerator` interface, which defines a single method:
+
+```php
+public function attempt(ComparisonNode $comparison) : ?Specification;
+```
+
+This is very similar to `BaseCompiler::generate()`, however returning a `Specification` is optional.
+
+Next, register an instance of your class with the `DelegatingCompiler`:
+
+```php
+$generator = new EmailAddressGenerator();
+$compiler  = new DelegatingCompiler($generator);
+```
+
+When `DelegatingCompiler::compile()` is invoked, the `DelegatingCompiler` will loop through all registered generators
+and call `SpecificationGenerator::attempt()` with each `ComparisonNode`.
+
+All `SpecificationGenerator`s provided via the `DelegatingCompiler`'s constructor share the same priority of `0`,
+however they can also be registered with an explicit priority:
+
+```php
+$generator = new EmailAddressGenerator();
+$compiler  = new DelegatingCompiler();
+
+$compiler->register($generator, 100); // Priority of 100.
+```
+
+`SpecificationGenerator`s with higher priority are invoked first.
+
+## Negating comparisons
+
+`ComparisonNode` does not expose negated comparisons like `does not equal` and `does not match`. However this is
+supported in the language by adding `not` before the comparison operator:
+
+```
+email not is "karl.rixon@gmail.com"
```

Comment:
s: Should we support negated comparisons like `does not equal`. It's quite hard to rationalise a statement like `email not is "karl.rixon@gmail.com"`

Reply:
You can use `email not ...` to make that case a bit more reasonable. It might be nice to also support `is not` (i.e. allowing the `not` to be before or after the operator). There will still be some weird cases like `not matches`/`matches not`, though we could define nicer aliases for these things (e.g. `misses` in the case of `matches`. Just depends how complicated we want to make the parser really.